src/peer_select.cc

```diff
@@ -99,6 +101,129 @@ operator <<(std::ostream &os, const PeerSelectionDumper &fsd)
+    void abortWaiting();
+    void noteWaitOver();
+
+    WaitingPeerSelectors selectors; ///< \see WaitingPeerSelectors
+};
+
+/// monitors all PeerSelector ICP ping timeouts
+static PeerSelectorPingMonitor &
+PingMonitor()
+{
+    static const auto Instance = new PeerSelectorPingMonitor();
+    return *Instance;
+}
+
+/* PeerSelectorPingMonitor */
+
+/// PeerSelectorPingMonitor::noteWaitOver() wrapper
+void
+PeerSelectorPingMonitor::NoteWaitOver(void *raw)
+{
+    assert(raw);
+    static_cast<PeerSelectorPingMonitor*>(raw)->noteWaitOver();
+}
+
+/// schedules a single event to represent all waiting selectors
+void
+PeerSelectorPingMonitor::startWaiting()
+{
+    assert(!selectors.empty());
+    const auto interval = tvSubDsec(current_time, selectors.begin()->first);
+    eventAdd("PeerSelectorPingMonitor::NoteWaitOver", &PeerSelectorPingMonitor::NoteWaitOver, this, interval, 0, false);
+}
+
+/// undoes an earlier startWaiting() call
+void
+PeerSelectorPingMonitor::abortWaiting()
+{
+    // our event may be already in the AsyncCallQueue but that is OK:
+    // such queued calls cannot accumulate, and we ignore any stale ones
+    eventDelete(&PeerSelectorPingMonitor::NoteWaitOver, nullptr);
+}
+
+/// calls back all ready PeerSelectors and continues to wait for others
+void
+PeerSelectorPingMonitor::noteWaitOver()
+{
+    while (!selectors.empty() && current_time >= selectors.begin()->first) {
+        const auto selector = selectors.begin()->second;
+        CallBack(selector->al, [selector,this] {
+            selector->ping.monitorRegistration = npos();
+            AsyncCall::Pointer callback = asyncCall(44, 4, "PeerSelector::HandlePingTimeout",
+                cbdataDialer(PeerSelector::HandlePingTimeout, selector));
+            ScheduleCallHere(callback);
+        });
+        selectors.erase(selectors.begin());
+    }
+
+    if (!selectors.empty()) {
+        // Since abortWaiting() is unreliable, we may have been awakened by a
+        // stale event A after event B has been scheduled. Now we are going to
+        // schedule event C. Prevent event accumulation by deleting B (if any).
+        abortWaiting();
+
+        startWaiting();
+    }
+}
+
+void
+PeerSelectorPingMonitor::monitor(PeerSelector *selector)
+{
+    assert(selector);
+
+    const auto deadline = selector->ping.deadline();
+    const auto position = selectors.emplace(deadline, selector);
+    selector->ping.monitorRegistration = position;
+
+    if (position == selectors.begin()) {
+        if (selectors.size() > 1)
+            abortWaiting(); // remove the previously scheduled earlier event
+        startWaiting();
+    } // else the already scheduled event is still the earliest one
+}
+
+void
+PeerSelectorPingMonitor::forget(PeerSelector *selector)
+{
+    assert(selector);
+
+    if (selector->ping.monitorRegistration == npos())
+        return; // already forgotten
+
+    const auto wasFirst = selector->ping.monitorRegistration == selectors.begin();
```

Comment:
> It should not matter whether this selector was first or not.

It maters if one wants to avoid creating stale events. If a non-first selector is removed, then there is no need to call abortWaiting() and startWaiting(). No stale event is created when a non-first selector is removed.

> The set is sorted by next pending timeout, there may be multiple entries still present with same timeout as the first one

True. We could detect the "last earliest" selector removal instead of the "first earliest" (for the lack of a better description). That check would be a bit more complex, but it would also optimize things further (in some use cases), so if it would address your concern, we should add it.

> the noteWaitOver() will efficiently re-schedule without sending notices if it gets called when no timeout is needed.

Will noteWaitOver() correctly reschedule upon being woken up on a stale event? Yes. Is that efficient? IMO, no.

IMO, an "efficient" implementation would not needlessly create stale events. An event shortens I/O wait and wakes up Squid (an expensive operation!). A stale event does that needlessly. A stale event may also cause questions in triage; it will even be visible to admins looking at the event queue. 

> But otherwise I think it better to do nothing with the event queue here.

I do not understand: Why is creating stale events better than avoiding them (to the extent possible) when avoiding them is simple? The implemented check itself is trivial. The "only the first event may change anything" logic is straightforward. Why is it better to ignore the problem of creating stale events?

> As optimization you could maybe call abortWaiting() when the selectors set becomes completely empty

Sure, but why stop there? The check for "empty" and the check for "first" cost about the same while the latter avoids stale events.

Reply:
>  We could detect the "last earliest" selector removal instead of the "first earliest" (for the lack of a better description).

I am not against this optimization. @yadij , will this change address your concern?

Reply:
That would, yes.

Reply:
Done at 2231d4.